torch/csrc/jit/codegen/cuda/lower_predicate.cpp

```diff
@@ -50,9 +50,73 @@ class ConditionalFromPredicateModifier : public kir::IrVisitor {
       setWritePredicate(expr, conditional);
     }
 
+    // This is special handling for async copy from gmem to smem.
+    //  The actual copy part is async to the gpu threads so we
+    //   cannot have both the copy and the initialization as we
+    //   initialize for other ops. This current approach inverts
```

Comment:
Today for vectorized support the pattern is:
Initialize buffer -> predicated load

For memcpy async:
If we initialized and then loaded (without sync) it would be undefined behavior...
1. Initialize only the "virtual out of boundary" accesses. Memory allocated, but outside the virtual tensor space. Virtual tensor space today is effectively what would be allocated in global memory.
2. Then only copy the "within bound" accesses

This is a WAR today based on how our system is set up. We would want to have a separate concept of SMEM space from Virtual or GMEM space, so that we know we're only working with the allocated SMEM. If we hit outside the allocated SMEM bad things happen.

Today asserting in predicate removal making sure that the virtual and SMEM boundaries line up based on the IterDomains.

Reply:
Added to comment.